### PR TITLE
misc: avoid splitting node labels in the middle of unicode surrogate pairs

### PR DESCRIPTION
**Pull request by wildlyinaccurate**
**Summary**

Potential fix for #11697. Takes advantage of the fact that the ES6 string iterator is unicode-aware.

**Related Issues/PRs**

fix for part of #11697

### CODE REVIEW

File: lighthouse-core/test/lib/page-functions-test.js [hunk 1]
```diff
@@ -116,6 +116,12 @@ describe('Page Functions', () => {
       assert.equal(pageFunctions.getNodeLabel(el).length, 80);
     });
 
+    it('Truncates long text containing unicode surrogate pairs', () => {
+      const el = dom.createElement('div');
+      el.innerText = Array(78).fill('a').join('') + '💡💡💡';
```
Comment (patrickhulce) [inline lighthouse-core/test/lib/page-functions-test.js:121]:
```suggestion
      // `getNodeLabel` truncates to 80 characters internally. 
      // We want to test a unicode character on the boundary.
      el.innerText = Array(78).fill('a').join('') + '💡💡💡';
```

**Review by patrickhulce — APPROVED** [review]
thanks very much @wildlyinaccurate! 🎉 

great find, PR much appreciated :)

**Comment by wildlyinaccurate** [comment]
Thanks for the suggestions @patrickhulce - much appreciated!

**Comment by wildlyinaccurate** [comment]
`yarn update:test-devtools` is failing on my machine. I'm not sure if I need to manually update some of the expected output fixtures or not?

**Comment by patrickhulce** [comment]
It's actually failing on master too @wildlyinaccurate , so not your doing. We'll let this chill for a bit until we're back up and running 👍

**Comment by connorjclark** [comment]
master on my machine: 
image

**Comment by patrickhulce** [comment]
@connorjclark it's in CI too https://github.com/GoogleChrome/lighthouse/runs/1434349934#step:15:206

**Comment by connorjclark** [comment]
Just landed #11702 to fix master.

How is `yarn update:test-devtools` failing, can you share the output? FYI it only works on Mac at the moment.

**Comment by patrickhulce** [comment]
FWIW I don't think you'll need to run `yarn update:test-devtools` once this is rebased though

**Comment by connorjclark** [comment]
I think the other 2 usages of `.slice` here are susceptible to the same issue. Hard to analyze the entire codebase tho, would be nice if one could search for `[String].slice` using the typescript parser.

**Comment by patrickhulce** [comment]
> I think the other 2 usages of .slice here are susceptible to the same issue. Hard to analyze the entire codebase tho, would be nice if one could search for [String].slice using the typescript parser.

There are many usages of `.slice` across LH that are susceptible to this issue. How do you feel about just leaving open #11697 for investigation by the team but still landing this PR?

**Comment by wildlyinaccurate** [comment]
> I think the other 2 usages of `.slice` here are susceptible to the same issue. Hard to analyze the entire codebase tho, would be nice if one could search for `[String].slice` using the typescript parser.

The attribute value truncation should probably use the `truncate` function. I guess the HTML snippet truncation could too, although `truncate` would need to be modified to take a custom truncation suffix.

**Comment by brendankenny** [comment]
> Hard to analyze the entire codebase tho, would be nice if one could search for [String].slice using the typescript parser.

find a `string.slice()`, right click, "Find All References"

**Comment by patrickhulce** [comment]
@googlebot I consent.

**Comment by patrickhulce** [comment]
@paulirish it looks like the buildtracker shared history doesn't work for external contributors, are you cool to disable it or temporarily revert?

File: .github/scripts/git-get-shared-history.sh [hunk 2]
```diff
@@ -20,6 +20,13 @@ set -euxo pipefail
 
 # We can always use some more history
 git -c protocol.version=2 fetch --deepen=100
+
+# Find out if the PR is coming from a fork
```
Comment (patrickhulce) [inline .github/scripts/git-get-shared-history.sh:24]: @paulirish how do you feel about this skipping this check entirely if it's from a fork for now instead and we can sort out the fix later?

File: .github/scripts/git-get-shared-history.sh [hunk 3]
```diff
@@ -20,6 +20,13 @@ set -euxo pipefail
...
+base_clone_url=$(jq '.pull_request.base.repo.clone_url' $GITHUB_EVENT_PATH)
```
Comment (patrickhulce) [inline .github/scripts/git-get-shared-history.sh:25]: shouldn't this be 
```suggestion
base_clone_url=$(jq '.pull_request.head.repo.clone_url' $GITHUB_EVENT_PATH)
```

to find out if it's a fork?

**Comment by wildlyinaccurate** [comment]
Hey, uh, does it help if I resubmit this against master as a single commit?

**Comment by paulirish** [comment]
> Hey, uh, does it help if I resubmit this against master as a single commit?

haha nah man it's all good. we'll just all pretend I never had to make all that noise.  the bug i had is fixed (as of my last commit), but it ends up not mattering. 

image

This should be good to go once the `yarn install` flake timeout is gone.

File: .github/scripts/git-get-shared-history.sh (hunk 2 above)
Comment (paulirish) [reply .github/scripts/git-get-shared-history.sh:24]: that works for me.. though determining that would probably have required almost this much iteration as well...    so i think it's all moot.

ALL MOOT!

**Review by paulirish — APPROVED** [review]
@wildlyinaccurate btw this is a fantastic find and fix. Nice work. :D

**Pull request merged**